### DISCUSSION TITLE
Retry and Rate Limit chaining

### DISCUSSION THREAD

**Discussion by LBalsa** [App-vNext/Polly#1718] [Q&A] [answered]
Hello,

I'm looking for clarification on whether retry and rate limiting can be chained together, due to the difference between v7 FILO and v8 FIFO. I need to wrap an external call which can return a number of exceptions with a retry, but also apply a proactive rate limit.

Will
`builder.AddRetry(...).AddRateLimiter(...).` 
Retry the execution on both the limiters' `RateLimiterRejectedException ` and any other exception, or should it be the other way round?
From my testing FIFO means these can't be used like this.
Alternatively, should I create two pipelines and call the rate limiter from within the retry one?
<img width="499" alt="image" src="https://github.com/App-vNext/Polly/assets/43744128/137bbf31-c270-4af6-bc35-5b5534c6b38d">

Thanks in advance

**Comment by peter-csala** [comment, marked as answer]
Hi LBalsa, the pollydocs have been recently extended with [several sequence diagrams regarding `ResiliencePipeline`s](https://www.pollydocs.org/pipelines/index.html#diagrams) to clarify how strategy chaining works.

> `builder.AddRetry(...).AddRateLimiter(...).`

In this case the retry will be your outer strategy and the rate limiter will be the inner. Depending on your rate limiter setup it either allows the request, queues it or rejects it. The retry will be triggered only if the rate limiter reject the request or the rate limiter does not handle the thrown exception.

> should I create two pipelines and call the rate limiter from within the retry one?

Nesting `ExecuteAsync` calls are considered anti-pattern. [Here](https://www.pollydocs.org/strategies/fallback.html#3---nesting-executeasync-calls) you can read more about this topic.
